src/main/java/com/limechain/trie/structure/BranchNodeHandle.java

```diff
@@ -0,0 +1,41 @@
+package com.limechain.trie.structure;
+
+/**
+ * Has the semantics of a "branch node with no storage value",
+ * i.e. in the purely "structural necessity branch node" sense.
+ */
+public final class BranchNodeHandle<T> extends NodeHandle<T> {
+
+    // NOTE: An ugly workaround to simulate the notion of "consuming" this branch node handle
+    private boolean consumed;
+    BranchNodeHandle(TrieStructure<T> trieStructure, int nodeIndex) {
+        super(trieStructure, nodeIndex);
+        consumed = false;
+    }
+
+    /**
+     * Marks this node as a storage node, thus converting {@code this} into a {@link StorageNodeHandle}.
+     * <br>
+     * This doesn't change the trie structure.
+     * @return the newly created {@link StorageNodeHandle}
+     * @implNote After calling this method, {@code this} instance is no longer valid,
+     *           since the underlying node has been converted to a storage node,
+     *           thus it's no longer a branch node.
+     *           Any further usage of this {@link BranchNodeHandle} would be invalid.
+     * @throws IllegalStateException if this branch node handle has already been consumed,
+     *                               i.e. the referenced node has been converted to a storage node.
+     */
+    public StorageNodeHandle<T> convertToStorageNode() {
```

Comment:
That's a direct issue when designing Java code influenced by Rust's tooling. Consider whether that's a compromise we're willing to allow (refer to the javadoc). :)

Reply:
Seems fine to me.

Reply:
> Seems fine to me.

Okay, in this case if we agree it's ok that's the approach I'm taking for `Vacant` as well. I'm awaiting final confirmation, still.

Reply:
Confirmation given

Reply:
Okay, so the case for `Vacant` turns out to be a rather more complicated one, since to invalidate the instance itself, it'll have to happen from within `PrepareInsert.One/.Two`. I feel this kinda stinks, so I'd suggest we make ourselves happy with only the `@implNote`. On the other hand, if we really want to be defensive about this, it's going to take a little more reference cluttering, which might damage readability... I'm not sure it's worth it, since such cases are really intentional misuse. 😄